File: src/apps/daemon/internal/networking/dnstunnel/hardware_advisor_test.go
package dnstunnel

import (
	"math"
	"testing"
)

func TestAdviseClientMobilePoorLossy(t *testing.T) {
	p := ClientHardwareProfile{
		Connection:           ConnMobile4g5g,
		Quality:              QualityPoor,
		CPUCores:             2,
		RAMMB:                2048,
		ResolverCensus:       CensusFew,
		MtuPreference:        MtuPrefStable,
		UseCase:              UseCaseChat,
		EncryptionPreference: EncLight,
	}

	rec := AdviseClient(p)

	if rec.DataEncryptionMethod != 1 {
		t.Errorf("expected XOR (1), got %d", rec.DataEncryptionMethod)
	}
	if rec.PacketDuplicationCount != 4 {
		t.Errorf("expected dup 4 for poor lossy, got %d", rec.PacketDuplicationCount)
	}
	if rec.SetupPacketDuplicationCount != 5 {
		t.Errorf("expected setup dup 5, got %d", rec.SetupPacketDuplicationCount)
	}
	if rec.ResolverBalancingStrategy != 3 {
		t.Errorf("expected strat 3 (lowest loss), got %d", rec.ResolverBalancingStrategy)
	}
	if rec.StreamResolverFailoverThreshold != 2 || rec.StreamResolverFailoverCooldownSec != 4 {
		t.Errorf("unexpected failover parameters: thresh %d, cool %d",
			rec.StreamResolverFailoverThreshold, rec.StreamResolverFailoverCooldownSec)
	}
	if rec.MinUploadMTU != 25 || rec.MaxUploadMTU != 60 {
		t.Errorf("unexpected stable MTU: min %d, max %d", rec.MinUploadMTU, rec.MaxUploadMTU)
	}
	if rec.UploadCompressionType != 1 {
		t.Errorf("expected ZSTD (1), got %d", rec.UploadCompressionType)
	}
	if rec.ArqWindowSize != 1000 || rec.ArqDataNackMaxGap != 64 || rec.ArqMaxDataRetries != 1500 {
		t.Errorf("unexpected lossy ARQ values")
	}
	if math.Abs(rec.DispatcherIdlePollIntervalSec-0.030) > 1e-6 {
		t.Errorf("expected 30ms poll for 2 cores, got %f", rec.DispatcherIdlePollIntervalSec)
	}
}

func TestAdviseClientFiberStreamStrong(t *testing.T) {
	p := ClientHardwareProfile{
		Connection:           ConnFiber,
		Quality:              QualityExcellent,
		CPUCores:             8,
		RAMMB:                8192,
		ResolverCensus:       CensusMany,
		MtuPreference:        MtuPrefSpeed,
		UseCase:              UseCaseStream,
		EncryptionPreference: EncStrong,
	}

	rec := AdviseClient(p)

	if rec.DataEncryptionMethod != 5 {
		t.Errorf("expected AES-256-GCM (5), got %d", rec.DataEncryptionMethod)
	}
	if rec.PacketDuplicationCount != 1 {
		t.Errorf("expected dup 1 for fiber excellent, got %d", rec.PacketDuplicationCount)
	}
	if rec.ResolverBalancingStrategy != 4 {
		t.Errorf("expected strat 4 (lowest latency), got %d", rec.ResolverBalancingStrategy)
	}
	if rec.MtuTestParallelism != 64 {
		t.Errorf("expected parallelism 64, got %d", rec.MtuTestParallelism)
	}
	if rec.UploadCompressionType != 2 {
		t.Errorf("expected LZ4 (2) for stream, got %d", rec.UploadCompressionType)
	}
	if rec.TxChannelSize != 16384 || rec.RxChannelSize != 16384 {
		t.Errorf("expected 16384 buffer sizes, got tx %d rx %d", rec.TxChannelSize, rec.RxChannelSize)
	}
	if math.Abs(rec.DispatcherIdlePollIntervalSec-0.010) > 1e-6 {
		t.Errorf("expected 10ms poll for stream, got %f", rec.DispatcherIdlePollIntervalSec)
	}
	if math.Abs(rec.PingAggressiveIntervalSec-0.15) > 1e-6 {
		t.Errorf("expected 0.15s aggressive ping, got %f", rec.PingAggressiveIntervalSec)
	}
}

func TestAdviseServerMinimalVsEnterprise(t *testing.T) {
	minP := ServerHardwareProfile{
		CPUCores:             1,
		RAMMB:                512,
		NetworkMbps:          100,
		ConcurrentUserCount:  1,
		Traffic:              TrafficLight,
		EncryptionPreference: EncLight,
		UpstreamDNS:          UpstreamCloudflare,
		LossyClients:         false,
	}
	minRec := AdviseServer(minP)
	if minRec.DataEncryptionMethod != 1 {
		t.Errorf("expected encryption 1, got %d", minRec.DataEncryptionMethod)
	}
	if minRec.UDPReaders != 2 || minRec.DNSRequestWorkers != 4 {
		t.Errorf("unexpected worker count for minimal server")
	}
	if minRec.MaxConcurrentRequests != 4096 {
		t.Errorf("expected 4096 max concurrent, got %d", minRec.MaxConcurrentRequests)
	}
	if minRec.SocketBufferSizeBytes != 4*1024*1024 {
		t.Errorf("expected 4MB socket buffer, got %d", minRec.SocketBufferSizeBytes)
	}
	if minRec.PacketBlockControlDuplication != 1 {
		t.Errorf("expected ctrl dup 1, got %d", minRec.PacketBlockControlDuplication)
	}

	entP := ServerHardwareProfile{
		CPUCores:             8,
		RAMMB:                4096,
		NetworkMbps:          1000,
		ConcurrentUserCount:  30,
		Traffic:              TrafficHeavy,
		EncryptionPreference: EncBalanced,
		UpstreamDNS:          UpstreamCombined,
		LossyClients:         true,
	}
	entRec := AdviseServer(entP)
	if entRec.DataEncryptionMethod != 2 {
		t.Errorf("expected encryption 2, got %d", entRec.DataEncryptionMethod)
	}
	if len(entRec.DNSUpstreamServers) != 4 {
		t.Errorf("expected 4 upstream servers, got %d", len(entRec.DNSUpstreamServers))
	}
	if entRec.UDPReaders != 4 || entRec.DNSRequestWorkers != 8 {
		t.Errorf("unexpected worker count for enterprise server")
	}
	if entRec.MaxConcurrentRequests != 32768 {
		t.Errorf("expected 32768 max concurrent, got %d", entRec.MaxConcurrentRequests)
	}
	if entRec.SocketBufferSizeBytes != 16*1024*1024 {
		t.Errorf("expected 16MB socket buffer, got %d", entRec.SocketBufferSizeBytes)
	}
	if entRec.DNSCacheMaxRecords != 100000 {
		t.Errorf("expected 100k cache records, got %d", entRec.DNSCacheMaxRecords)
	}
	if entRec.PacketBlockControlDuplication != 3 || entRec.MaxPacketsPerBatch != 12 {
		t.Errorf("expected ctrl dup 3, batch 12 for lossy clients")
	}
	if entRec.SessionTimeoutSeconds != 180 {
		t.Errorf("expected 180s timeout, got %d", entRec.SessionTimeoutSeconds)
	}
	if entRec.SocksConnectTimeoutSeconds != 60 {
		t.Errorf("expected 60s connect timeout, got %d", entRec.SocksConnectTimeoutSeconds)
	}
	if entRec.SessionCleanupIntervalSeconds != 15 {
		t.Errorf("expected 15s cleanup interval, got %d", entRec.SessionCleanupIntervalSeconds)
	}
	if entRec.Socks5FragmentStoreCapacity != 2048 {
		t.Errorf("expected 2048 fragment store capacity, got %d", entRec.Socks5FragmentStoreCapacity)
	}
}

File: src/packages/lumicore/tests/hardware_advisor_test.rs
use lumicore::dns::hardware_advisor::*;

#[test]
fn test_client_advice_mobile_poor_lossy() {
    let profile = ClientHardwareProfile {
        connection: NetworkConnection::Mobile4g5g,
        quality: NetworkQuality::Poor,
        cpu_cores: 2,
        ram_mb: 2048,
        resolver_census: ResolverCensus::Few,
        mtu_preference: MtuPreference::Stable,
        use_case: UseCase::Chat,
        encryption_preference: EncryptionPreference::Light,
    };

    let rec = HardwareAdvisor::advise_client(&profile);

    // Light encryption -> XOR (1)
    assert_eq!(rec.data_encryption_method, 1);
    // Poor quality -> lossy -> duplication = 4
    assert_eq!(rec.packet_duplication_count, 4);
    assert_eq!(rec.setup_packet_duplication_count, 5);
    // Lossy -> strategy 3 (Lowest loss)
    assert_eq!(rec.resolver_balancing_strategy, 3);
    // Lossy failover
    assert_eq!(rec.stream_resolver_failover_resend_threshold, 2);
    assert_eq!(rec.stream_resolver_failover_cooldown_sec, 4);
    // Stable MTU
    assert_eq!(rec.min_upload_mtu, 25);
    assert_eq!(rec.max_upload_mtu, 60);
    // Mobile/lossy chat compression -> ZSTD (1)
    assert_eq!(rec.upload_compression_type, 1);
    // Lossy ARQ
    assert_eq!(rec.arq_window_size, 1000);
    assert_eq!(rec.arq_data_nack_max_gap, 64);
    assert_eq!(rec.arq_max_data_retries, 1500);
    // 2 CPU cores poll interval -> 30ms
    assert!((rec.dispatcher_idle_poll_interval_sec - 0.030).abs() < f32::EPSILON);
}

#[test]
fn test_client_advice_fiber_stream_strong() {
    let profile = ClientHardwareProfile {
        connection: NetworkConnection::Fiber,
        quality: NetworkQuality::Excellent,
        cpu_cores: 8,
        ram_mb: 8192,
        resolver_census: ResolverCensus::Many,
        mtu_preference: MtuPreference::Speed,
        use_case: UseCase::Stream,
        encryption_preference: EncryptionPreference::Strong,
    };

    let rec = HardwareAdvisor::advise_client(&profile);

    // Strong encryption -> AES-256-GCM (5)
    assert_eq!(rec.data_encryption_method, 5);
    // Fiber excellent -> duplication = 1
    assert_eq!(rec.packet_duplication_count, 1);
    assert_eq!(rec.setup_packet_duplication_count, 2);
    // Many resolvers non-lossy -> strategy 4 (Lowest latency)
    assert_eq!(rec.resolver_balancing_strategy, 4);
    // Speed MTU
    assert_eq!(rec.min_upload_mtu, 60);
    assert_eq!(rec.max_download_mtu, 700);
    // Parallelism with 8 cores and Many resolvers -> 64
    assert_eq!(rec.mtu_test_parallelism, 64);
    // Stream + High RAM -> LZ4 compression (2), 16384 buffers
    assert_eq!(rec.upload_compression_type, 2);
    assert_eq!(rec.tx_channel_size, 16384);
    assert_eq!(rec.rx_channel_size, 16384);
    // Stream poll interval -> 10ms
    assert!((rec.dispatcher_idle_poll_interval_sec - 0.010).abs() < f32::EPSILON);
    // Aggressive ping for stream
    assert!((rec.ping_aggressive_interval_sec - 0.15).abs() < f32::EPSILON);
}

#[test]
fn test_server_advice_minimal_vs_enterprise() {
    // Minimal single user server
    let min_prof = ServerHardwareProfile {
        cpu_cores: 1,
        ram_mb: 512,
        network_mbps: 100,
        concurrent_users: 1,
        traffic: ServerTraffic::Light,
        encryption_preference: EncryptionPreference::Light,
        upstream_dns: ServerDnsUpstream::Cloudflare,
        clients_lossy: false,
    };
    let min_rec = HardwareAdvisor::advise_server(&min_prof);
    assert_eq!(min_rec.data_encryption_method, 1);
    assert_eq!(min_rec.udp_readers, 2);
    assert_eq!(min_rec.dns_request_workers, 4);
    assert_eq!(min_rec.max_concurrent_requests, 4096);
    assert_eq!(min_rec.socket_buffer_size_bytes, 4 * 1024 * 1024);
    assert_eq!(min_rec.dns_cache_max_records, 10_000);
    assert_eq!(min_rec.packet_block_control_duplication, 1);
    assert_eq!(min_rec.session_timeout_seconds, 300);

    // Enterprise 30+ users heavy server with lossy clients
    let ent_prof = ServerHardwareProfile {
        cpu_cores: 8,
        ram_mb: 4096,
        network_mbps: 1000,
        concurrent_users: 30,
        traffic: ServerTraffic::Heavy,
        encryption_preference: EncryptionPreference::Balanced,
        upstream_dns: ServerDnsUpstream::Combined,
        clients_lossy: true,
    };
    let ent_rec = HardwareAdvisor::advise_server(&ent_prof);
    assert_eq!(ent_rec.data_encryption_method, 2);
    assert_eq!(ent_rec.dns_upstream_servers.len(), 4);
    assert_eq!(ent_rec.udp_readers, 4);
    assert_eq!(ent_rec.dns_request_workers, 8);
    assert_eq!(ent_rec.max_concurrent_requests, 32768);
    assert_eq!(ent_rec.socket_buffer_size_bytes, 16 * 1024 * 1024);
    assert_eq!(ent_rec.dns_cache_max_records, 100_000);
    assert_eq!(ent_rec.packet_block_control_duplication, 3);
    assert_eq!(ent_rec.max_packets_per_batch, 12);
    assert_eq!(ent_rec.session_timeout_seconds, 180);
    assert_eq!(ent_rec.socks_connect_timeout_seconds, 60);
    assert_eq!(ent_rec.session_cleanup_interval_seconds, 15);
    assert_eq!(ent_rec.socks5_fragment_store_capacity, 2048);
}
